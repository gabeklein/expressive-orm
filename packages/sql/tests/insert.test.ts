import { Type, Num, Str } from '..';
import { inMemoryDatabase } from '.';

class Foo extends Type {
  name = Str();
  color = Str();
}

class User extends Type  {
  name = Str();
  email = Str();
  age = Num();
}

it.skip("will insert procedurally generated rows", async () => {
  await inMemoryDatabase([User]);

  const names = [
    "Gabe",
    "Justin",
    "Joe",
    "Steven",
    "Josiah"
  ];

  await User.insert(
    names.map((name, i) => ({
      name,
      age: i + 25,
      email: `${name.toLowerCase()}@email.org`
    }))
  );

  // const results = await Query.get(where => where(User))

  // void results;

  // const authors: any[] = [];
  // const books: any[] = [];

  // for(const authorName of ["Gabe", "BJ", "JC", "Danny"]){
  //   const author = new Author();

  //   author.name = authorName;
  //   authors.push(author);

  //   for(const name of ["Foo", "Bar", "Baz"]){
  //     const book = new Book();

  //     book.title = name;
  //     book.author = author;
      
  //     books.push(book);
  //   }
  // }

  // connection.insert([ ...authors, ...books ])
})

it("will throw for bad value", async () => {
  const insert = () => (
    Foo.insert({
      name: "foobar",
      // @ts-expect-error
      color: 3
    }) 
  )

  expect(insert).toThrowErrorMatchingSnapshot();
})

it("will throw for no value non-nullable", async () => {
  const insert = () => (
    // TODO: This should also have a type warning.
    //// @ts-expect-error
    Foo.insert({
      name: "foobar"
    }) 
  )

  expect(insert).toThrowErrorMatchingSnapshot();
})

it("will add index to specify error", async () => {
  const insert = () => (
    Foo.insert([
      {
        name: "foo",
        color: "red"
      },
      // TODO: This should also have a type warning.
      //// @ts-expect-error
      {
        name: "bar"
      }
    ]) 
  )

  expect(insert).toThrowErrorMatchingSnapshot();
})

it.todo("will warn in typescript for bad value");
it.todo("will warn in typescript for no value non-nullable");