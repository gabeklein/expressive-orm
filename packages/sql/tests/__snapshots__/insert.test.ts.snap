// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will add index to specify error 1`] = `Provided value for Foo.color is undefined on index [1] but column is not nullable.`;

exports[`will throw for bad value 1`] = `
Provided value for Foo.color but not acceptable for type VARCHAR(255).
Value must be a string.
`;

exports[`will throw for no value non-nullable 1`] = `Provided value for Foo.color is undefined but column is not nullable.`;
